pkg/reconciler/pipelinerun/resources/validate_params_test.go

```diff
@@ -121,3 +121,83 @@ func TestValidateParamTypesMatching_Invalid(t *testing.T) {
 		})
 	}
 }
+
+func TestValidateExpectedParametersProvided_Valid(t *testing.T) {
+	tcs := []struct {
+		name          string
+		p             *v1beta1.Pipeline
+		pr            *v1beta1.PipelineRun
+		errorExpected bool
+	}{{
+		name: "expected params provided",
+		p: tb.Pipeline("a-pipeline", tb.PipelineSpec(
+			tb.PipelineParamSpec("expected-string-param", v1beta1.ParamTypeString),
+			tb.PipelineParamSpec("expected-array-param", v1beta1.ParamTypeArray))),
+		pr: tb.PipelineRun("a-pipelinerun", tb.PipelineRunSpec("test-pipeline",
+			tb.PipelineRunParam("expected-string-param", "somestring"),
+			tb.PipelineRunParam("expected-array-param", "another", "array"))),
+		errorExpected: false,
+	}, {
+		name:          "no params to get wrong",
+		p:             tb.Pipeline("a-pipeline"),
+		pr:            tb.PipelineRun("a-pipelinerun"),
+		errorExpected: false,
+	}, {
+		name: "expected string param missing",
+		p: tb.Pipeline("a-pipeline", tb.PipelineSpec(
+			tb.PipelineParamSpec("expected-string-param", v1beta1.ParamTypeString),
+			tb.PipelineParamSpec("expected-array-param", v1beta1.ParamTypeArray))),
+		pr: tb.PipelineRun("a-pipelinerun", tb.PipelineRunSpec("test-pipeline",
+			tb.PipelineRunParam("expected-array-param", "another", "array"))),
+		errorExpected: true,
+	}, {
+		name: "expected array param missing",
+		p: tb.Pipeline("a-pipeline", tb.PipelineSpec(
+			tb.PipelineParamSpec("expected-string-param", v1beta1.ParamTypeString),
+			tb.PipelineParamSpec("expected-array-param", v1beta1.ParamTypeArray))),
+		pr: tb.PipelineRun("a-pipelinerun", tb.PipelineRunSpec("test-pipeline",
+			tb.PipelineRunParam("expected-string-param", "somestring"))),
+		errorExpected: true,
+	}}
+	for _, tc := range tcs {
+		t.Run(tc.name, func(t *testing.T) {
+			err := ValidateExpectedParametersProvided(&tc.p.Spec, tc.pr)
+			if (!tc.errorExpected) && (err != nil) {
+				t.Errorf("Pipeline.Validate() returned error: %v", err)
+			}
+
+			if tc.errorExpected && (err == nil) {
+				t.Error("Pipeline.Validate() did not return error, wanted error")
+			}
+		})
+	}
+}
+
+func TestValidateExpectedParametersProvided_Invalid(t *testing.T) {
+	tcs := []struct {
+		name string
+		p    *v1beta1.Pipeline
+		pr   *v1beta1.PipelineRun
```

Comment:
i think unit tests are a great tool to use to guide writing cleaner code and i think there is an interesting signal in this test: to setup the test, you have to create entire pipelines and pipelienruns, though all you really care about is the params.

it would be an easier test to read and maintain if you only had to have lists of the params required and the params provided.

what would it take to make it so that ValidateExpectedParametersProvided only takes the lists of params as arguments, instead of the entire pipeline and pipelinerun?